File: ports/csharp/src/network/kernel/BcsSocket.cs
using System;
using System.Net.Sockets;
using Bcs.Core.Kernel;

namespace Bcs.Network.Kernel
{
    public class BcsTcpSocket : BcsObject
    {
        private TcpClient _client;
        private NetworkStream _stream;
        private readonly object _lock = new object();

        public BcsTcpSocket(BcsObject parent = null) : base(parent)
        {
        }

        public void ConnectToHost(string address, int port)
        {
            lock (_lock)
            {
                _client?.Close();
                _client = new TcpClient();
                _client.Connect(address, port);
                _stream = _client.GetStream();
            }
        }

        public int Write(byte[] data)
        {
            lock (_lock)
            {
                if (_stream == null || !_client.Connected)
                    throw new InvalidOperationException("Socket not connected.");

                _stream.Write(data, 0, data.Length);
                return data.Length;
            }
        }

        public int Read(byte[] data)
        {
            lock (_lock)
            {
                if (_stream == null || !_client.Connected)
                    throw new InvalidOperationException("Socket not connected.");

                return _stream.Read(data, 0, data.Length);
            }
        }

        public void Close()
        {
            lock (_lock)
            {
                _stream?.Close();
                _client?.Close();
                _stream = null;
                _client = null;
            }
        }
    }
}

File: ports/rust/src/network/kernel/bcs_socket.rs
use std::sync::{Arc, Mutex};
use std::net::TcpStream;
use std::io::{Read, Write};
use crate::core::kernel::bcs_object::BcsObject;

pub struct BcsTcpSocket {
    pub base: Arc<BcsObject>,
    stream: Mutex<Option<TcpStream>>,
}

impl BcsTcpSocket {
    pub fn new(parent: Option<Arc<BcsObject>>) -> Arc<Self> {
        let base = if let Some(p) = parent {
            BcsObject::new_with_parent(p)
        } else {
            BcsObject::new()
        };

        Arc::new(Self {
            base,
            stream: Mutex::new(None),
        })
    }

    pub fn connect_to_host(&self, address: &str, port: u16) -> std::io::Result<()> {
        let mut stream_guard = self.stream.lock().unwrap();
        let addr = format!("{}:{}", address, port);
        let stream = TcpStream::connect(addr)?;
        *stream_guard = Some(stream);
        Ok(())
    }

    pub fn write(&self, data: &[u8]) -> std::io::Result<usize> {
        let mut stream_guard = self.stream.lock().unwrap();
        if let Some(stream) = stream_guard.as_mut() {
            stream.write(data)
        } else {
            Err(std::io::Error::new(std::io::ErrorKind::NotConnected, "Socket not connected"))
        }
    }

    pub fn read(&self, data: &mut [u8]) -> std::io::Result<usize> {
        let mut stream_guard = self.stream.lock().unwrap();
        if let Some(stream) = stream_guard.as_mut() {
            stream.read(data)
        } else {
            Err(std::io::Error::new(std::io::ErrorKind::NotConnected, "Socket not connected"))
        }
    }

    pub fn close(&self) {
        let mut stream_guard = self.stream.lock().unwrap();
        *stream_guard = None;
    }
}
